### DISCUSSION TITLE
Json serialization, annotations, and openapi swagger

### DISCUSSION THREAD

**Discussion by CharlesPerrotMinot** [pydantic/pydantic#10227] [Question] [answered]
Hi!

Moving from Pydantic V1 to V2, to support the serialization of all `datetime` and `Decimal` objects, we used the following in our  Model class (that inherits from BaseModel and that we use as our own flavoured BaseModel):
```py
@field_serializer("*", mode="wrap")
def serialize_special_types(
    self,
    value: Any,
    nxt: SerializerFunctionWrapHandler,
    _info: SerializationInfo,
) -> Any:
    if isinstance(value, datetime):
        return isoformat(value)
    if isinstance(value, Decimal):
        return str(value)
    return nxt(value)
```
- Using an `annotated` type isn't really an option as we have many models, and ensuring all never use the special type instead of the builtin one is not practically doable nor maintainable.
- We have way too many models to make a serializer for ever field.

If you know of any better way to do this, please share!

The problem? It breaks the openapi.json swagger. The only way to fix it was to remove the return annotation (`Any` in the example) otherwise all the types of any class using the Model with that `field_serializer` are lost in the openapi swagger.

For reference, here is the openapi swagger with the Any annotation:
<img width="321" alt="Screenshot 2024-08-24 at 10 41 35 PM" src="https://github.com/user-attachments/assets/18fb8256-a904-45f4-9432-330d9e050fa5">

That same model in the openapi swagger if the return annotation for the `field_serializer` is removed:
<img width="540" alt="Screenshot 2024-08-24 at 10 39 57 PM" src="https://github.com/user-attachments/assets/2db6ad0e-de1c-4fba-b3b9-63e7f36649d3">

Is there any other way to deal with this?
We have mypy and ruff checkers, which means we have to annotate the function twice for this use-case `# type: ignore[no-untyped-def]` and `# noqa: ANN201`

Thanks! 

PS: we are using the latest pydantic (2.8.2) and FastAPI (0.112.2) versions

**Comment by Viicos** [comment]
The best solution would be to leave the method return type unannotated, but indeed this can trigger linter warnings. Maybe a sentinel value could be introduced and special cased in Pydantic, but this seems really specific. I'll discuss this with the team tomorrow.

**Comment by Viicos** [comment]
Considering `@field_serializer` applied to all fields (with `*`) are pretty uncommon, and that you can still ignore the issue with specific comments, we are pretty reluctant to support this. This can be implemented with the following:

```py
class NoInferSentinel:  # Name TBD.
    """A sentinel value indicating that even if the return type annotations is present, it shouldn't be used."""
    ...

@field_serializer("*", mode="wrap", return_type=NoInferSentinel)
def serialize_special_types(
    self,
    value: Any,
    nxt: SerializerFunctionWrapHandler,
    _info: SerializationInfo,
) -> Any:
    if isinstance(value, datetime):
        return isoformat(value)
    if isinstance(value, Decimal):
        return str(value)
    return nxt(value)
```

Happy to review a PR, although I still feel like introducing a new member to the code base for this use case is unfortunate.

**Reply by CharlesPerrotMinot** [reply]
As much as I agree with your conclusion (being reluctant to support an uncommon case), I'm wondering if the use case I'm presenting is that uncommon. Are you expecting everyone to use annotated types? Or serializers for every field by name? That seems pretty unlikely for people having spread-out models in different apps, with different attribute names. What would you recommend as being the go-to?

Based on documentation: "_Pydantic uses the terms "serialize" and "dump" interchangeably. Both refer to the process of converting a model to a dictionary or JSON-encoded string._"
As a result, wouldn't it make sense for the openapi.json to be based on the model annotation and not the serializer annotation? Since in the end, in order to send the response, most data ends up getting string-encoded anyways? To be true to the actual return, I feel like `@field_serializer("*")` should actually be annotated with a `str` return.

**Reply by Viicos** [reply, marked as answer]
> As much as I agree with your conclusion (being reluctant to support an uncommon case), I'm wondering if the use case I'm presenting is that uncommon. Are you expecting everyone to use annotated types? Or serializers for every field by name?

I don't exactly know your use case, but it seems you could define reusable annotated types:

```py
MyDatetime = Annotated[datetime, PlainSerializer(isoformat)]

# Then use `MyDatetime` instead of `datetime` in model annotations
```

This avoids the need to define a `field_serializer('*')` for _every_ model. What do you think?

---

> wouldn't it make sense for the openapi.json to be based on the model annotation and not the serializer annotation?

If your serializer returns a `str` for a type annotated as `int`, it needs to be reflected accordingly.

> To be true to the actual return, I feel like `@field_serializer("*")` should actually be annotated with a `str` return.

Well this isn't necessarily true, and `Any` is more correct than `str` (that may be correct in most cases, unless the return type is a boolean, a number, null, etc).

As said, if you want to stick with the decorator approach, `-> Any` can be omitted. Having to add ignore comments can be annoying, but it is only once per model.

**Reply by CharlesPerrotMinot** [reply]
So, we use our own "flavoured" base model, that inherits from the pydantic BaseModel, where we implement a couple of things, including this serializer, a hash calculation, a modified _freeze to support list / set / dict, and a json dump that support all data types (UUID, enum, dates, ...) ; though some may now be irrelevant with pydantic v2 (still have to double check that, they were needed with v1).

Long story short, we use it as our own base model, and all our models inherit from it. I think using the Annotated type like you mention works when you have a few models in a couple of apps, but typically we have hundreds of Models across more than 50 applications, and not only migrating to the new types, but also ensuring that all models never use the builtin ones, seems pretty tedious if even doable.

Good points about the annotations! Thanks!
